watertap/unit_models/pressure_changer.py

```diff
@@ -11,35 +11,301 @@
+    **VariableEfficiency.none** - uses default pump efficiency at BEP
+    **VariableEfficiency.flow** - uses an efficiency correlation scaled to the BEP flow rate
+    **VariableEfficiency.flow_head** - uses an efficiency correlation scaled to the BEP flow rate and head
+    """,
+        ),
+    )
+
+    def build(self):
+        super().build()
+
+        # ---------------------------------------
+        # Isothermal pump set up
+        # ---------------------------------------
+
+        self.control_volume.del_component(self.control_volume.enthalpy_balances)
+
+        @self.control_volume.Constraint(
+            self.flowsheet().config.time, doc="Isothermal constraint"
+        )
+        def isothermal_balance(b, t):
+            return b.properties_in[t].temperature == b.properties_out[t].temperature
+
+        # ---------------------------------------
+        # Variable efficiency pump set-up
+        # ---------------------------------------
+
+        if self.config.variable_efficiency is not VariableEfficiency.none:
+            # create additional pyomo variables
+            self.bep_flow = Var(
+                initialize=1.0,
+                doc="Best efficiency point flowrate of the centrifugal pump",
+                units=pyunits.m**3 / pyunits.s,
+            )
+
+            self.bep_eta = Var(
+                initialize=self.efficiency_pump[0].value,
+                doc="Best efficiency of the centrifugal pump",
+                units=pyunits.dimensionless,
+            )
+
+            self.flow_ratio = Var(
+                self.flowsheet().time,
+                initialize=1.0,
+                doc="Ratio of pump flowrate to best efficiency point flowrate",
+                units=pyunits.dimensionless,
+            )
+
+            # unfix the default efficiency from the base class
+            self.efficiency_pump.unfix()
+
+            # add constraints
+            @self.Constraint(self.flowsheet().time, doc="Pump flow ratio")
+            def flow_ratio_constraint(b, t):
+                return b.flow_ratio[t] * b.bep_flow == (
+                    b.control_volume.properties_out[t].flow_vol.expr()
+                )
+
+        if self.config.variable_efficiency is VariableEfficiency.flow:
+
+            @self.Expression(
+                self.flowsheet().time,
+                doc="Expression for variable pump efficiency based on flow only",
+            )
+            def eta_ratio(b, t):
+                return Expr_if(
+                    b.flow_ratio[t] < 0.6,
+                    0.4,
+                    Expr_if(
+                        b.flow_ratio[t] > 1.4,
+                        0.4,
+                        -0.995 * b.flow_ratio[t] ** 2 + 1.977 * b.flow_ratio[t] + 0.018,
+                    ),
+                )
+
+        elif self.config.variable_efficiency is VariableEfficiency.flow_head:
+            raise ValueError(
```

Comment:
I think we may have a `NotImplemented` exception although in the past, others have recommended not using it since users may expect to have such features readily available. In this case though, I don't see a problem with using the NotImplemented exception.